CONTRIBUTING.md

```diff
@@ -23,66 +23,48 @@ maximize the chances of your PR being merged.
+versioning guidelines:
+
+* Features may be marked as deprecated in a given versioned API at any point in time, but this may
+  only be done when a replacement implementation and configuration path is available in Envoy on
+  master. The litmus test is to imagine that a stateless translation tool exists that
+  could convert from the earlier API to the new API. A field may be deprecated
+  if this tool would be able to perform the conversion. For example, removing a
+  field to describe HTTP/2 window settings is valid if a more comprehensive
+  HTTP/2 protocol options field is being introduced to replace it. The PR author
+  deprecating the old configuration is responsible for updating all tests and
+  canonical configuration, or guarding them with the `DEPRECATED_FEATURE_TEST()` macro.
+  This will be validated by the `bazel.compile_time_options` target, which will hard-fail when
+  deprecated configuration is used.
+* We reserve the right to delete deprecated configuration across major API versions. E.g. a field
+  marked deprecated in v2 may be removed in v3.
 * Unless the community and Envoy maintainer team agrees on an exception, during the
   first release cycle after a feature has been deprecated, use of that feature
   will cause a logged warning, and incrementing the
   [runtime](https://www.envoyproxy.io/docs/envoy/latest/configuration/runtime#config-runtime)
-  runtime.deprecated_feature_use stat.
+  `runtime.deprecated_feature_use` stat.
   During the second release cycle, use of the deprecated configuration will
   cause a configuration load failure, unless the feature in question is
   explicitly overridden in
   [runtime](https://www.envoyproxy.io/docs/envoy/latest/configuration/runtime#config-runtime)
-  config. Finally during the third release cycle the code and configuration will be removed
-  entirely.
+  config. Finally, following the deprecation of the API major version where the field was first
+  marked deprecated, the entire implementation code will be removed from the Envoy implementation.
 * This policy means that organizations deploying master should have some time to get ready for
-  breaking changes, but we make no guarantees about the length of time.
+  breaking changes at the next major API version. This is typically a window of at least 12 months
+  or until the organization moves to the next major API version.
 * The breaking change policy also applies to source level extensions (e.g., filters). Code that
   conforms to the public interface documentation should continue to compile and work within the
   deprecation window. Within this window, a warning of deprecation should be carefully logged (some
   features might need rate limiting for logging this). We make no guarantees about code or deployments
   that rely on undocumented behavior.
 * All deprecations/breaking changes will be clearly listed in the [deprecated log](docs/root/intro/deprecated.rst).
```

Comment:
Can we flag this as process which should be changed?
Ideally if we're writing the tooling which converts old-foo to new-foo we can autogenerate the docs.  Tracking bug for Q4 perhaps?

Reply:
Tracked in https://github.com/envoyproxy/envoy/issues/8465